### PR TITLE
Introduce go chi web framework as frontend of macaron, so that we can move routes from macaron to chi step by step

### PR DESCRIPTION
**Pull request by lunny**
According to #7403, we should chose go-chi as Gitea's backend web framework.

Why we should move away from macaron, see @zeripath 's comment https://github.com/go-gitea/gitea/issues/7403#issuecomment-509728528

This PR will not finish all the work to remove macaron, but it will introduce go chi as the frontend of macaron. Only most middlewares and serval routes will be handled by go chi, and most routes will still be handled by macaron. To implement that, a go chi middleware is written to delegate the HTTP requests to macaron.

A difference between chi and macaron is macaron will set req.Body back but chi will not after consumed the request. That's why the changes to the test is in this PR.

It's ready to review now.

### CODE REVIEW

**Comment by lunny** [comment]
related with #7403

**Review by 6543 — APPROVED** [review]
its gitea_v1.14 time ... lests start with it

**Review by lafriks — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -39,6 +39,7 @@ require (
 	github.com/facebookgo/subset v0.0.0-20150612182917-8dac2c3c4870 // indirect
 	github.com/gliderlabs/ssh v0.3.1
 	github.com/glycerine/go-unsnap-stream v0.0.0-20190901134440-81cf024a9e0a // indirect
+	github.com/go-chi/chi v4.1.2+incompatible
```
Comment (lafriks) [inline go.mod]: Just saw that go mod based go-chi v1.5.0 has just been released, we should use it
```suggestion
	github.com/go-chi/chi v1.5.0
```

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@lafriks done.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@lunny the chi router do not pass things to our logger :/

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
:rocket:

**Comment by lafriks** [comment]
LGTM does not seems to like me :/

**Pull request merged**